src/lib/opts.rs

```diff
@@ -363,20 +366,209 @@ pub fn setup() -> Opts {
+    /// FASTQ files are gathered using the path prefix(es), with their kind (e.g. sample barcode or
+    /// template), kind number (e.g. read one or index two), and lane inferred from the file name.
+    /// Reads across differnt lanes with the same kind and kind number are grouped, and a
+    /// corresponding read structure is built assuming all the bases in that read are of the given
+    /// kind.  The groups are then sorted by kind and kind number to proper ordering of sample
+    /// barcodes, UMIs, and template reads (i.e. read pairs).
+    ///
+    /// If the input FASTQS are a mix of files and path prefixes, this function will return an
+    /// error.
+    pub fn from(opts: &Opts) -> Result<Vec<FastqsAndReadStructure>> {
+        // If the input FASTQs are in fact path prefixes, then glob in the FASTQs and create the
+        // read structure based on the kind/kind-number inferred from the file name.
+        let input_fastq_group: Vec<FastqsAndReadStructure> = if opts
+            .fastqs
+            .iter()
+            .all(|f| f.is_file())
+        {
+            // must have FASTQs
+            ensure!(!opts.fastqs.is_empty(), "No FASTQs found with --fastq");
+
+            // must have read structures, and match the # of FASTQs
+            ensure!(
+                opts.fastqs.len() == opts.read_structures.len(),
+                "Same number of read structures should be given as FASTQs"
+            );
+
+            opts.fastqs
+                .iter()
+                .zip(opts.read_structures.iter())
+                .map(|(f, r)| FastqsAndReadStructure {
+                    fastqs: vec![f.clone()],
+                    read_structure: r.clone(),
+                })
+                .collect()
+        } else if opts.fastqs.iter().all(|f| !f.is_file()) {
+            ensure!(
+                opts.read_structures.is_empty(),
+                "Read Structure must not be given when the input FASTQs are a path prefix."
+            );
+
+            // glob all the FASTQs
+            let input_fastqs: Vec<InputFastq> =
+                opts.fastqs.iter().flat_map(|prefix| InputFastq::glob(&prefix).unwrap()).collect();
+
+            // must find some FASTQs
+            ensure!(!input_fastqs.is_empty(), "No FASTQS found for prefix(es): {:?}.", opts.fastqs);
+
+            // The next few steps (1) group FASTQS by kind and kind number, (2) re-order the groups
+            // by the kind and kind number, and (3) sorts the FASTQS in each group by prefix and
+            // lane.  (1) the group of FASTQS with the same kind and kind number must be read in
+            // serially, and (3) ensures that the FASTQS are synchronized across groups (e.g. the
+            // same lane/prefix across kind/kind-number will be read at the same time).  (2)
+            // ensures that we emit the read pairs in the correct order (e.g. R1/R2).
+
+            // Collect all FASTQs by kind and kind number
```

Comment:
nb: `group_by` only groups _consecutive elements, not across the entire vector/iterator. Hence why I'm inserting into a hash map here.